### DISCUSSION TITLE
pcf8574 used for input

### DISCUSSION THREAD

**Discussion by eichenwind** [micropython/micropython#10380] [RP2040 / Pico]
Hi, I looked at the code of https://github.com/mcauser/micropython-pcf8574 and as far as I read the pcf8574 can be used as input or output device. In the code I see that the functions state and direction as described in https://micropython-pcf8574.readthedocs.io/en/latest/pcf8574.html are not used. So I added those functions in the code and I'm able to read them. But it seems that this has no impact to the chip.

so I added in the class:

class PCF8574:
    def __init__(self, i2c, address=0x20):
        self._i2c = i2c
        self._address = address
        self._direction = 0xFF
        self_state = 0xFF
        self._port = bytearray(1)
        if i2c.scan().count(address) == 0:
            raise OSError('PCF8574 not found at I2C address {:#x}'.format(address)) 

This should define all ports as Input ports, seems not to work

If I set direction to 0x0 and state to direction to 0x0, then all ports should be off. Seems also not to work. Does anybody knows where I make a mistake?

**Comment by davefes** [comment]
Looking at the error and the linked docs have you grounded A0, A1 and A2?

> So I added those functions in the code and I'm able to read them.

What does "I'm able to read them" mean?

**Reply by eichenwind** [reply]
> Looking at the error and the linked docs have you grounded A0, A1 and A2?

yes

> > So I added those functions in the code and I'm able to read them.
> What does "I'm able to read them" mean?

print("direction:", pcf._direction)
print("state",pcf._state)

**Comment by scruss** [comment]
It's not clear to me from your question, but your readthedocs link is for a different library than mcauser's one. Does the [psolyca/micropython-pcf8574](https://github.com/psolyca/micropython-pcf8574) library not work for you?

I've successfully read switch inputs using mcauser's library before

**Reply by rolandvs** [reply]
The chip has 4 registers:
Base address: 0x20 or 0x38 depending on the chip suffix (nothing or A).
The chip default is all inputs.
So if you want to read a switch: use a pullup of 10k and a switch to ground, the center to one of the pins.
Read the base address multiple times and see if you can notice the difference when pressing and releasing the switch.

This can all be done using some commands on REPL without worrying about a driver and give you better understanding of the problem.

**Reply by selamu033** [reply]
Can you please tell me on which page of the PCF8574 datasheet I find the four registers. I checked the datasheet but can't find the 4 registers.

**Reply by rkompass** [reply]
[TIs datasheet](https://www.ti.com/lit/ds/symlink/pcf8574.pdf?ts=1733670017232&ref_url=https%253A%252F%252Fwww.ti.com%252Fproduct%252Fde-de%252FPCF8574) tells that there are no such registers. On page 12:

>  This device does not have internal configuration or
status registers. Instead, read or write to the device I/Os directly after sending the device address (see Figure
7-3 and Figure 7-4).
